Add noverlap layout algorithm

apitts opened the pull request:
This layout algorithm is a port of the noverlap layout algorithm from Gephi (with the generous permission of @jacomyma). It is useful to remove the overlap between nodes and also to generally improve the appearance of networks by adding configurable margins between nodes. It is relevant to a number of issues, including:
https://github.com/jacomyal/sigma.js/issues/687
https://github.com/jacomyal/sigma.js/issues/107
https://github.com/jacomyal/sigma.js/issues/613

Yomguithereal commented:
Hello @apitts. I'll start reviewing your PR.

Yomguithereal commented:
Is this `noverlap` or `noOverlap` then?

Yomguithereal commented:
I notice that your documentation lacks the `stop` and `kill` method.

Yomguithereal commented on plugins/sigma.layout.noverlap/README.md:
````diff
@@ -0,0 +1,80 @@
+Starts the layout. It is possible to pass a configuration if this is the first time you start the layout.
+
+```js
+var listener = sigma.layout.noverlap.start(sigInst, config);
````
Why don't you attach methods to sigma's prototype like FA2 does?

Yomguithereal commented on plugins/sigma.layout.noverlap/sigma.layout.noverlap.js:
```diff
@@ -0,0 +1,408 @@
+
+      adjacentNodes = {}; //An object that stores the node ids of adjacent nodes (either in same grid box or adjacent grid box) for all nodes
+
+      for(row = 0; row < self.config.gridSize; row++) {
```
Five nested loops? This might get costly very fast no?

apitts replied:
Do you think so? I aimed to be as close as possible as the original algorithm when porting. Clearly two of the nested loops are only ever of size 3 (plus and minus one on each side for adjacent cells). Of course, if you have any performance improvements in mind that would be great. I was thinking it may be possible to update the bounds every x iterations say instead of on every iteration...but I'm not certain about how that would impact convergence and I suspect it wouldn't have a massive performance gain. For reasonable values of margin I've found the algorithm quite performant on reasonable sized graphs and you can always increase grid size for larger graphs to get more performance. Of course, convergence is much tougher where the margin is large.

Yomguithereal replied:
Can you test it on a large graph (say at least ~5000 nodes) and tell me whether this works?

apitts commented:
Re the naming. I called it "noverlap" to be consistent with the name in Gephi (single o, not double o). I'm not wedded to it though. I was initially calling the layout Distribute Nodes which seems a bit more descriptive to me as it does more than just eliminate overlap. Happy to change it to that. 

I will add documentation on stop and kill methods.

Re attaching methods to sigma prototype like FA2. I used another layout (dagre) as inspiration and have a slight preference for keeping the methods aligned to that layout. I think of FA2 as a different beast given it uses webworkers. But if you feel strongly about it, I can take another look.

Yomguithereal commented:
> Re the naming. I called it "noverlap" to be consistent with the name in Gephi (single o, not double o).

Fair enough :)

> Re attaching methods to sigma prototype like FA2. I used another layout (dagre) as inspiration and have a slight preference for keeping the methods aligned to that layout. I think of FA2 as a different beast given it uses webworkers. But if you feel strongly about it, I can take another look.

The only issue here is consistency. It might be weird to ask users to do two different things when using different layout when they are basically the same.

apitts commented:
@Yomguithereal I have made the following changes:

1. Changed the methods so that they attach to sigma.prototype like FA2 and updated example and docs accordingly. 
2. Changed the example to 5,000 nodes and included a timing report for the algorithm. Clearly there are a lot of moving parts in terms of run time - it's possible to cause a lot of collisions by making nodes very large or demanding extreme margins between nodes. It's also possible to find an optimal gridSize which will depend on the degree to which nodes overlap and how many of them there are...I have increased gridSize here but haven't attempted to optimise. It's a random network so run time will vary but I'm showing between 1.5 seconds and 2.5 seconds on my desktop for this 5,000 node network. That seems OK to me...most use cases should involve much less overlap than this random network. I anticipate most people using the layout after running FA2, potentially after FA2 has been run with adjustSizes at the end (which I think should be updated by the way to use rendered node sizes as per this layout). 

Re the documentation on kill and stop...I actually don't think they are public methods (i.e. not part of the interface) so hence not documented. Let me know if you disagree.

Yomguithereal commented:
> Re the documentation on kill and stop...I actually don't think they are public methods (i.e. not part of the interface) so hence not documented. Let me know if you disagree.

If the layout stop itself and clean its own allocated memory then ok.

I'll review the changes tomorrow. Don't hesitate to ping if I forget to do so.

Yomguithereal commented:
So, is everything ok here? Can I merge?

apitts commented:
Yep - I think good to go. The layout stops itself and cleans its own allocated memory. Thanks.

Yomguithereal commented:
Thanks @apitts. Just merged!
